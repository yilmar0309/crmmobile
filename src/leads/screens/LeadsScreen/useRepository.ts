import {mapLeadsData} from '@/leads/data/mappers/mappers';
import {LeadsModel} from '@/leads/data/models/leadsModel';
import {
  useGetAllLeadsQuery,
  useValidateLeadsMutation,
} from '@/leads/data/remote/leadsApi';

export function useRepository(): {
  leadsData: LeadsModel[];
  isFetching: boolean;
  handleValidateLeadApi: (id: string) => {
    unwrap(): Promise<unknown>;
  };
  refetch: () => void;
} {
  const {data, isFetching, refetch} = useGetAllLeadsQuery('');
  const [handleValidateLeadApi] = useValidateLeadsMutation();

  const leadsData = data ? data.map(mapLeadsData) : [];

  return {
    leadsData,
    isFetching,
    refetch,
    handleValidateLeadApi,
  };
}
