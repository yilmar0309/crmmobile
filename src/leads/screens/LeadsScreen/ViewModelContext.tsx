import React, {createContext, ReactNode, useCallback, useContext} from 'react';
import Toast from 'react-native-toast-message';
import {LeadsEntity} from '@/leads/data/remote/entities/leadsEntity';
import {useRepository} from './useRepository';

type ViewModelContextType = {
  leadsData: LeadsEntity[];
  isFetching: boolean;
  getItem: (dataItem: LeadsEntity[], index: number) => LeadsEntity;
  getItemCount: () => number;
  handleValidateLead: (id: number) => Promise<void>;
  refetch: () => void;
};

const ViewModelContext = createContext<ViewModelContextType | undefined>(
  undefined,
);

export function ViewModelProvider({children}: {children: ReactNode}) {
  const {leadsData, isFetching, refetch, handleValidateLeadApi} =
    useRepository();

  const handleValidateLead = useCallback(
    async (id: number) => {
      try {
        await handleValidateLeadApi(id.toString());
        Toast.show({
          type: 'success',
          position: 'top',
          text1: 'Validation success',
        });
      } catch (error) {
        Toast.show({
          type: 'error',
          position: 'top',
          text1: 'Validation Failed',
          text2: 'There was an error validating this lead.',
        });
      }
    },
    [handleValidateLeadApi],
  );

  const getItem = (dataItem: LeadsEntity[], index: number): LeadsEntity =>
    dataItem[index];

  const getItemCount = () => leadsData.length;

  return (
    <ViewModelContext.Provider
      value={{
        leadsData,
        isFetching,
        refetch,
        getItem,
        getItemCount,
        handleValidateLead,
      }}>
      {children}
    </ViewModelContext.Provider>
  );
}

export function useViewModelProvider() {
  const context = useContext(ViewModelContext);
  if (context === undefined) {
    throw new Error('Publication View Model Provider');
  }
  return context;
}
