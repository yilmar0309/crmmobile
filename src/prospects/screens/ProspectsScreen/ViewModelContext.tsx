import React, {createContext, ReactNode, useContext} from 'react';
import {ProspectsEntity} from '@/prospects/data/remote/entities/prospectsEntity';
import {useRepository} from './useRepository';

type ViewModelContextType = {
  isFetching: boolean;
  prospectsData: ProspectsEntity[];
  getItem: (dataItem: ProspectsEntity[], index: number) => ProspectsEntity;
  getItemCount: () => number;
  refetch: () => void;
};

const ViewModelContext = createContext<ViewModelContextType | undefined>(
  undefined,
);

export function ViewModelProvider({children}: {children: ReactNode}) {
  const {prospectsData, isFetching, refetch} = useRepository();

  const getItem = (
    dataItem: ProspectsEntity[],
    index: number,
  ): ProspectsEntity => dataItem[index];

  const getItemCount = () => prospectsData.length;

  return (
    <ViewModelContext.Provider
      value={{
        isFetching,
        prospectsData,
        getItem,
        getItemCount,
        refetch,
      }}>
      {children}
    </ViewModelContext.Provider>
  );
}

export function useViewModelProvider() {
  const context = useContext(ViewModelContext);
  if (context === undefined) {
    throw new Error('Publication View Model Provider');
  }
  return context;
}
