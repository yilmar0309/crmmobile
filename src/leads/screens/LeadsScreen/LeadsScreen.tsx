import React from 'react';
import {VirtualizedList} from 'react-native';
import {AppContainer, AppLoadingList} from '@/core/components';
import {AppRenderCard} from '@/core/components/AppRenderCard';
import {TabsHomeScreenProps} from '@/core/types/StackRoutes';
import {LeadsEntity} from '@/leads/data/remote/entities/leadsEntity';
import {LeadsStackRoutes} from '@/leads/routes';
import {useViewModelProvider, ViewModelProvider} from './ViewModelContext';

const LeadsScreenContent =
  ({}: TabsHomeScreenProps<LeadsStackRoutes.LEADS>) => {
    const {
      leadsData,
      isFetching,
      refetch,
      getItem,
      getItemCount,
      handleValidateLead,
    } = useViewModelProvider();

    return (
      <AppContainer>
        {isFetching ? (
          <AppLoadingList />
        ) : (
          <VirtualizedList
            testID="virtualized-list"
            data={leadsData}
            renderItem={({item}: {item: LeadsEntity}) => (
              <AppRenderCard item={item} onValidate={handleValidateLead} />
            )}
            getItemCount={getItemCount}
            getItem={getItem}
            keyExtractor={item => item.id.toString()}
            refreshing={isFetching}
            onRefresh={refetch}
          />
        )}
      </AppContainer>
    );
  };

export const LeadsScreen = (
  props: TabsHomeScreenProps<LeadsStackRoutes.LEADS>,
) => (
  <ViewModelProvider>
    <LeadsScreenContent {...props} />
  </ViewModelProvider>
);
