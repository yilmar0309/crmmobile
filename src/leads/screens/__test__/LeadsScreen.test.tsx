// src/__tests__/LeadsScreen.test.tsx

import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import {LeadsScreen} from '../LeadsScreen';
import * as leadsViewModel from '../LeadsScreen/ViewModelContext';

jest.mock('../LeadsScreen/ViewModelContext', () => ({
  useViewModelProvider: jest.fn(),
  ViewModelProvider: ({children}: {children: React.ReactNode}) => children,
}));

jest.mock('@/core/components/AppRenderCard', () => {
  return {
    AppRenderCard: jest.fn(() => null),
  };
});

jest.mock('@/core/components', () => {
  const {View} = require('react-native');
  return {
    AppContainer: ({
      children,
      testID,
    }: {
      children: React.ReactNode;
      testID?: string;
    }) => <View testID={testID}>{children}</View>,
    AppLoadingList: ({testID}: {testID?: string}) => (
      <View testID={testID || 'app-loading-list'} />
    ),
  };
});

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...props,
});

describe('LeadsScreen', () => {
  let props: any;
  let AppRenderCardMock: jest.Mock;

  beforeEach(() => {
    props = createTestProps({});
    jest.clearAllMocks();
    AppRenderCardMock = require('@/core/components/AppRenderCard')
      .AppRenderCard as jest.Mock;
  });

  it('renders correctly when isFetching is false', () => {
    const mockLeadsData = [
      {
        id: 1,
        name: 'Lead 1',
      },
      {
        id: 2,
        name: 'Lead 2',
      },
    ];

    const mockViewModelReturn = {
      leadsData: mockLeadsData,
      isFetching: false,
      refetch: jest.fn(),
      getItem: (data: any, index: number) => data[index],
      getItemCount: () => mockLeadsData.length,
      handleValidateLead: jest.fn(),
    };

    (leadsViewModel.useViewModelProvider as jest.Mock).mockImplementation(
      () => mockViewModelReturn,
    );

    const {getByTestId, toJSON} = render(<LeadsScreen {...props} />);

    expect(getByTestId('virtualized-list')).toBeTruthy();

    expect(AppRenderCardMock).toHaveBeenCalledTimes(mockLeadsData.length);

    mockLeadsData.forEach((lead, index) => {
      expect(AppRenderCardMock).toHaveBeenNthCalledWith(
        index + 1,
        {
          item: lead,
          onValidate: mockViewModelReturn.handleValidateLead,
        },
        {},
      );
    });

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders loading state when isFetching is true', () => {
    const mockViewModelReturn = {
      leadsData: [],
      isFetching: true,
      refetch: jest.fn(),
      getItem: jest.fn(),
      getItemCount: jest.fn(),
      handleValidateLead: jest.fn(),
    };

    (leadsViewModel.useViewModelProvider as jest.Mock).mockImplementation(
      () => mockViewModelReturn,
    );

    const {getByTestId} = render(<LeadsScreen {...props} />);

    expect(getByTestId('app-loading-list')).toBeTruthy();
  });

  it('calls refetch when list is refreshed', () => {
    const mockLeadsData = [
      {
        id: 1,
        name: 'Lead 1',
      },
      {
        id: 2,
        name: 'Lead 2',
      },
    ];

    const mockRefetch = jest.fn();

    const mockViewModelReturn = {
      leadsData: mockLeadsData,
      isFetching: false,
      refetch: mockRefetch,
      getItem: (data: any, index: number) => data[index],
      getItemCount: () => mockLeadsData.length,
      handleValidateLead: jest.fn(),
    };

    (leadsViewModel.useViewModelProvider as jest.Mock).mockImplementation(
      () => mockViewModelReturn,
    );

    const {getByTestId} = render(<LeadsScreen {...props} />);

    const list = getByTestId('virtualized-list');

    fireEvent(list, 'onRefresh');

    expect(mockRefetch).toHaveBeenCalled();
  });
});
