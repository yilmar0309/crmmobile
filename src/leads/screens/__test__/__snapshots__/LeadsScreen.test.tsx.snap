// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeadsScreen renders correctly when isFetching is false 1`] = `
<View>
  <RCTScrollView
    data={
      [
        {
          "id": 1,
          "name": "Lead 1",
        },
        {
          "id": 2,
          "name": "Lead 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[MockFunction]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[MockFunction]}
        refreshing={false}
      />
    }
    refreshing={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    testID="virtualized-list"
  >
    <RCTRefreshControl />
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      />
    </View>
  </RCTScrollView>
</View>
`;
